File: query/pkg/model/query.go
package model

import "time"

// Query represents the canonical JSON query structure for TelHawk.
// All query interfaces (UI filter chips, text syntax, raw JSON) translate to this format.
type Query struct {
	Select       []string      `json:"select,omitempty"`
	Filter       *FilterExpr   `json:"filter,omitempty"`
	TimeRange    *TimeRangeDef `json:"timeRange,omitempty"`
	Aggregations []Aggregation `json:"aggregations,omitempty"`
	Sort         []SortSpec    `json:"sort,omitempty"`
	Limit        int           `json:"limit,omitempty"`
	Offset       int           `json:"offset,omitempty"`
	Cursor       string        `json:"cursor,omitempty"`
}

// FilterExpr represents a filter expression which can be either a simple condition
// or a compound condition (AND, OR, NOT).
type FilterExpr struct {
	// Simple condition fields
	Field    string      `json:"field,omitempty"`
	Operator string      `json:"operator,omitempty"`
	Value    interface{} `json:"value,omitempty"`

	// Compound condition fields
	Type       string       `json:"type,omitempty"` // "and", "or", "not"
	Conditions []FilterExpr `json:"conditions,omitempty"`
	Condition  *FilterExpr  `json:"condition,omitempty"` // for NOT
}

// TimeRangeDef defines the time bounds for a query.
// Supports both absolute timestamps and relative time ranges.
type TimeRangeDef struct {
	Start *time.Time `json:"start,omitempty"`
	End   *time.Time `json:"end,omitempty"`
	Last  string     `json:"last,omitempty"` // e.g., "15m", "1h", "24h", "7d"
}

// Aggregation defines a statistical aggregation to perform on query results.
type Aggregation struct {
	Type         string        `json:"type"`                   // "terms", "date_histogram", "avg", "sum", "min", "max", "stats", "cardinality"
	Field        string        `json:"field,omitempty"`        // OCSF field path (e.g., ".actor.user.name")
	Name         string        `json:"name"`                   // Identifier for this aggregation in results
	Size         int           `json:"size,omitempty"`         // For terms aggregations
	Interval     string        `json:"interval,omitempty"`     // For date_histogram (e.g., "1h")
	Aggregations []Aggregation `json:"aggregations,omitempty"` // Nested aggregations
}

// SortSpec defines a field to sort by and the sort direction.
type SortSpec struct {
	Field string `json:"field"`           // OCSF field path (e.g., ".time")
	Order string `json:"order,omitempty"` // "asc" or "desc"
}

// Supported filter operators
const (
	OpEq         = "eq"         // Equals
	OpNe         = "ne"         // Not equals
	OpGt         = "gt"         // Greater than
	OpGte        = "gte"        // Greater than or equal
	OpLt         = "lt"         // Less than
	OpLte        = "lte"        // Less than or equal
	OpIn         = "in"         // In array
	OpContains   = "contains"   // String contains
	OpStartsWith = "startsWith" // String starts with
	OpEndsWith   = "endsWith"   // String ends with
	OpRegex      = "regex"      // Regular expression
	OpExists     = "exists"     // Field exists
	OpCIDR       = "cidr"       // IP in CIDR range
)

// Supported compound filter types
const (
	FilterTypeAnd = "and"
	FilterTypeOr  = "or"
	FilterTypeNot = "not"
)

// Supported aggregation types
const (
	AggTypeTerms         = "terms"
	AggTypeDateHistogram = "date_histogram"
	AggTypeAvg           = "avg"
	AggTypeSum           = "sum"
	AggTypeMin           = "min"
	AggTypeMax           = "max"
	AggTypeStats         = "stats"
	AggTypeCardinality   = "cardinality"
)

// IsSimpleCondition returns true if this filter is a simple field condition (not compound).
func (f *FilterExpr) IsSimpleCondition() bool {
	return f.Type == "" && f.Field != ""
}

// IsCompoundCondition returns true if this filter is a compound condition (AND/OR/NOT).
func (f *FilterExpr) IsCompoundCondition() bool {
	return f.Type != ""
}

File: web/frontend/src/types/query.ts
/**
 * TelHawk JSON Query Language Types
 *
 * These types define the canonical JSON query structure used for all search,
 * filtering, and analysis operations across TelHawk.
 *
 * See: docs/QUERY_LANGUAGE_DESIGN.md
 */

/**
 * Supported query operators
 */
export type QueryOperator =
  | 'eq'           // Equals
  | 'ne'           // Not equals
  | 'gt'           // Greater than
  | 'gte'          // Greater than or equal
  | 'lt'           // Less than
  | 'lte'          // Less than or equal
  | 'in'           // In array
  | 'contains'     // String contains
  | 'startsWith'   // String starts with
  | 'endsWith'     // String ends with
  | 'regex'        // Regular expression
  | 'exists'       // Field exists
  | 'cidr';        // IP in CIDR range

/**
 * Basic filter condition (field-operator-value)
 */
export interface FilterCondition {
  field: string;
  operator: QueryOperator;
  value: string | number | boolean | string[] | number[];
}

/**
 * Compound filter with AND/OR/NOT logic
 */
export interface CompoundFilter {
  type: 'and' | 'or' | 'not';
  conditions?: FilterExpression[];  // For 'and' and 'or'
  condition?: FilterExpression;     // For 'not'
}

/**
 * A filter expression can be either a simple condition or a compound filter
 */
export type FilterExpression = FilterCondition | CompoundFilter;

/**
 * Type guard to check if a filter is a compound filter
 */
export function isCompoundFilter(filter: FilterExpression): filter is CompoundFilter {
  return 'type' in filter && ['and', 'or', 'not'].includes(filter.type);
}

/**
 * Type guard to check if a filter is a filter condition
 */
export function isFilterCondition(filter: FilterExpression): filter is FilterCondition {
  return 'field' in filter && 'operator' in filter && 'value' in filter;
}

/**
 * Time range specification
 */
export interface TimeRange {
  // Absolute time range
  start?: string;  // ISO 8601 timestamp
  end?: string;    // ISO 8601 timestamp

  // Relative time range
  last?: string;   // e.g., "15m", "1h", "24h", "7d", "30d", "90d"
}

/**
 * Aggregation types
 */
export type AggregationType =
  | 'terms'          // Group by discrete values
  | 'date_histogram' // Group by time buckets
  | 'avg'            // Average
  | 'sum'            // Sum
  | 'min'            // Minimum
  | 'max'            // Maximum
  | 'stats'          // All metrics (count, avg, sum, min, max)
  | 'cardinality';   // Unique count

/**
 * Aggregation specification
 */
export interface Aggregation {
  type: AggregationType;
  field: string;
  name: string;
  size?: number;              // For terms aggregation
  interval?: string;          // For date_histogram (e.g., "1h", "1d")
  aggregations?: Aggregation[]; // Nested aggregations
}

/**
 * Sort specification
 */
export interface SortSpec {
  field: string;
  order: 'asc' | 'desc';
}

/**
 * Complete JSON query structure
 */
export interface Query {
  select?: string[];           // Field projection (optional)
  filter?: FilterExpression;   // WHERE clause (optional)
  timeRange?: TimeRange;       // Time filtering (optional but recommended)
  aggregations?: Aggregation[]; // GROUP BY / stats (optional)
  sort?: SortSpec[];           // ORDER BY (optional)
  limit?: number;              // Result limit (default: 100)
  offset?: number;             // Pagination offset (optional)
  cursor?: string;             // Cursor-based pagination (optional)
}

/**
 * Query response structure
 */
export interface QueryResponse<T = any> {
  events?: T[];                // Search results
  total: number;               // Total matches
  cursor?: string;             // Next page cursor
  aggregations?: Record<string, AggregationResult>; // Aggregation results
  took?: number;               // Query duration in ms
}

/**
 * Aggregation result bucket
 */
export interface AggregationBucket {
  key: string | number;
  count: number;
  [key: string]: any; // Nested aggregations or metrics
}

/**
 * Aggregation result structure
 */
export interface AggregationResult {
  buckets?: AggregationBucket[];
  value?: number;  // For metric aggregations
  [key: string]: any; // Nested aggregations
}
